contributors/design-proposals/federated-statefulsets.md

```diff
@@ -0,0 +1,491 @@
+will be aptly modified to ensure that the dns records against the headless services are also updated
+in the federated dns server if the need be. The federation service controller will determine this based
+on the user annotation (```federation.kubernetes.io/statefulset-lb-needed```) with value ```true``` present
+on the federated headless service. Please not that this is the same annotation which is used on the
+federated statefulset to determine, if this statefulset needs LB services or not. The federation
+statefulset controller will update this on the federated headless service while creating it, or if
+its applied/updated later, will update it accordingly.
+Additionally the information about the stateful replicas (and the same name LB services) for which the
+service controller should update the dns records against the headless service will also need to be passed
+on to the headless service. The proposal is to pass this information through annotations on the headless
+service as below:
+```
+federation.kubernetes.io/stateful-replica-list: <replica-name-1>, <replica-name-2>, ..., <replica-name-n>
+```
+The federation service controller, if finds the names in this list against a federated headless service,
+will look for matching federated LB services and update the CNAME record for the headless service with the
+relevant SRV records for all the LB services in the federation dns server.
+This list can certainly be updated to the first class API data, if the federation API objects and clients
+are updated and maintained separately to have additional fields/data then that present in k8s API.
+
+Once this happens, the stateful pods can discover peers by querying the governing service domain (federated
+headless service). Taking the same example listed earlier if the peer (or a user) queries the service
+```store.mynamespace.myfederation.svc.example.com```, the below SRV records will be returned.
+```
+store-c1-0.etcd.mynamespace.myfederation.svc.example.com
+store-c1-1.etcd.mynamespace.myfederation.svc.example.com
+store-c1-2.etcd.mynamespace.myfederation.svc.example.com
+store-c1-3.etcd.mynamespace.myfederation.svc.example.com
+store-c2-0.etcd.mynamespace.myfederation.svc.example.com
+store-c2-1.etcd.mynamespace.myfederation.svc.example.com
+store-c2-2.etcd.mynamespace.myfederation.svc.example.com
+store-c2-3.etcd.mynamespace.myfederation.svc.example.com
+store-c3-0.etcd.mynamespace.myfederation.svc.example.com
+store-c3-1.etcd.mynamespace.myfederation.svc.example.com
+store-c3-2.etcd.mynamespace.myfederation.svc.example.com
+```
+The global federation dns service will have SRV records and relevant A records listing IPs of the LBs.
+The local in-cluster kube-dns will have the SRV records and the relevant A records listing the local
+stateful set pod IPs. The local kube-dns will be able to respond with both the local and global records,
+when a query hits the local kube-dns. The peers can choose to connect to the local identities or the
```

Comment:
I think its app specific and app decides, which identity to use. We just provide a feature where a federated app can have that facility.